File: services/realtime/internal/engine/nnue/features.go
// Package nnue is Phase 3 of the engine rebuild: a real quantized,
// incrementally-updated evaluation network, replacing engine/search's
// placeholder material+overlay eval.go the way the plan calls for.
//
// Scope, stated plainly: this is a genuine, working, tested
// accumulator-based quantized network with card-aware inputs -- not a
// production-grade, fully-tuned one. Real differences from a
// state-of-the-art NNUE (e.g. Stockfish's): a single shared accumulator
// computed from White's own king-bucket perspective, rather than a true
// dual White-sees/Black-sees accumulator pair; 4 coarse king buckets
// (quadrants) rather than per-square HalfKP buckets; one quantization
// scale shared across layers rather than per-layer tuning. These are
// deliberate simplifications to keep the whole pipeline (features,
// incremental accumulator, quantized inference, training, round-trip
// verification) tractable and CORRECT within this phase, rather than
// partially building a more ambitious architecture. Widening the king
// bucketing or adding a second perspective later is additive, not a
// rewrite, since the accumulator/quantization machinery doesn't care how
// many buckets or perspectives feed it.
package nnue

import (
	"github.com/chess404/realtime/internal/engine/actions"
	"github.com/chess404/realtime/internal/engine/core"
)

const (
	// numKingBuckets divides the board into quadrants by White's own king
	// position -- the "condition every feature on the king" idea HalfKP
	// captures precisely, simplified to 4 coarse buckets instead of 64.
	numKingBuckets = 4
	// numPieceSquareFeatures: 64 squares x 5 non-king piece types x 2
	// colors. Kings are excluded -- a king's own square is exactly what
	// determines the bucket, so encoding it again as a piece-square feature
	// would be redundant.
	numPieceSquareFeatures = 64 * 5 * 2
	numChessFeatures       = numKingBuckets * numPieceSquareFeatures

	// numCountBuckets buckets a count into {0, 1, 2-or-more} -- coarse on
	// purpose; a learned network doesn't need the exact count of frozen
	// pieces to know "having several is different from having one".
	numCountBuckets = 3
	// Card-aware auxiliary features (deliberately NOT king-bucketed --
	// these are global facts about the position, not spatial ones):
	// hand size (mine, theirs), frozen piece count (mine, theirs),
	// shielded piece count (mine, theirs) -- each a numCountBuckets-wide
	// one-hot -- plus a fortress-active boolean (mine, theirs).
	numCardCountFeatures = numCountBuckets * 6
	numCardBoolFeatures  = 2
	numCardFeatures      = numCardCountFeatures + numCardBoolFeatures

	// NumFeatures is the total input dimension: every accumulator row
	// (Network.L1Weights) has exactly this many entries.
	NumFeatures = numChessFeatures + numCardFeatures
)

var pieceTypeIndex = map[core.PieceType]int{
	core.Pawn: 0, core.Knight: 1, core.Bishop: 2, core.Rook: 3, core.Queen: 4,
}

// kingBucket buckets kingSq into one of the 4 board quadrants.
func kingBucket(kingSq core.Square) int {
	bucket := 0
	if kingSq.File() >= 4 {
		bucket |= 1
	}
	if kingSq.Rank() >= 4 {
		bucket |= 2
	}
	return bucket
}

// pieceSquareFeature is the (non-king-bucketed) index for a piece of type
// pt and color c sitting on sq -- combined with a king bucket by
// ActiveFeatures/KingBucketOf to form a full feature index.
func pieceSquareFeature(sq core.Square, pt core.PieceType, c core.Color) int {
	return int(sq)*5*2 + pieceTypeIndex[pt]*2 + int(c)
}

// KingBucketOf exposes kingBucket for callers that need to detect a
// king-bucket change (e.g. deciding whether an incremental update or a
// full Refresh is required after a move) without recomputing it themselves.
func KingBucketOf(p *core.Position) int {
	return kingBucket(p.KingSquare(core.White))
}

// PieceFeature returns the full (king-bucketed) feature index for a piece
// of type pt and color c on sq, given bucket (the current king bucket) --
// the incremental-update primitive: when a non-king piece moves, its OLD
// square's feature is removed and its NEW square's feature is added, both
// computed via this same function with the SAME bucket (since only the
// king's own move changes the bucket).
func PieceFeature(bucket int, sq core.Square, pt core.PieceType, c core.Color) int {
	return bucket*numPieceSquareFeatures + pieceSquareFeature(sq, pt, c)
}

func appendCountBucket(feats []int, base, count int) ([]int, int) {
	bucket := count
	if bucket > numCountBuckets-1 {
		bucket = numCountBuckets - 1
	}
	return append(feats, base+bucket), base + numCountBuckets
}

func countWhere(p *core.Position, c core.Color, match func(sq core.Square) bool) int {
	count := 0
	pieces := p.Occupied(c)
	for pieces.Any() {
		var sq core.Square
		sq, pieces = pieces.PopLSB()
		if match(sq) {
			count++
		}
	}
	return count
}

// ActiveFeatures returns every active feature index for (p, ov,
// whiteHand, blackHand) -- the full, from-scratch feature set a Refresh
// needs. Deliberately takes actions.Hand directly (not a bundled struct)
// so this package has no dependency on engine/search, which is the
// package that will import THIS one to use as an eval -- avoiding an
// import cycle.
func ActiveFeatures(p *core.Position, ov *core.CardOverlay, whiteHand, blackHand actions.Hand) []int {
	bucket := kingBucket(p.KingSquare(core.White))
	feats := make([]int, 0, 32)

	for sq := core.Square(0); sq < 64; sq++ {
		piece := p.PieceAt(sq)
		if piece.IsNone() || piece.Type == core.King {
			continue
		}
		feats = append(feats, PieceFeature(bucket, sq, piece.Type, piece.Color))
	}

	base := numChessFeatures
	feats, base = appendCountBucket(feats, base, len(whiteHand))
	feats, base = appendCountBucket(feats, base, len(blackHand))
	feats, base = appendCountBucket(feats, base, countWhere(p, core.White, ov.IsFrozen))
	feats, base = appendCountBucket(feats, base, countWhere(p, core.Black, ov.IsFrozen))
	feats, base = appendCountBucket(feats, base, countWhere(p, core.White, ov.IsShielded))
	feats, base = appendCountBucket(feats, base, countWhere(p, core.Black, ov.IsShielded))
	if ov.HasFortress(core.White) {
		feats = append(feats, base)
	}
	base++
	if ov.HasFortress(core.Black) {
		feats = append(feats, base)
	}
	return feats
}

File: services/realtime/internal/engine/nnue/pytrainer/features.py
"""Python port of internal/engine/nnue/features.go.

This module MUST stay index-for-index identical to the Go implementation.
test_roundtrip.py verifies that against testdata/golden_features.json,
which the Go test TestGenerateNNUEFeatureFixtures
(../fixtures_test.go) regenerates from the live Go code on every run --
so the golden file can never silently drift out of sync with Go, only
Python's port of it can, which is exactly what the round-trip test
catches.

Only piece placement and White's king square come from the FEN; every
other field (hand sizes, frozen/shielded counts, fortress booleans) is
passed in directly, matching ActiveFeatures' own signature (it takes
counts, not squares -- see that function's doc comment).
"""

NUM_KING_BUCKETS = 4
# 64 squares x 5 non-king piece types x 2 colors.
NUM_PIECE_SQUARE_FEATURES = 64 * 5 * 2
NUM_CHESS_FEATURES = NUM_KING_BUCKETS * NUM_PIECE_SQUARE_FEATURES

# Buckets a count into {0, 1, 2-or-more}.
NUM_COUNT_BUCKETS = 3
NUM_CARD_COUNT_FEATURES = NUM_COUNT_BUCKETS * 6
NUM_CARD_BOOL_FEATURES = 2
NUM_CARD_FEATURES = NUM_CARD_COUNT_FEATURES + NUM_CARD_BOOL_FEATURES

NUM_FEATURES = NUM_CHESS_FEATURES + NUM_CARD_FEATURES

WHITE, BLACK = 0, 1

# core.PieceType values (types.go): NoPieceType=0, Pawn=1, Knight=2,
# Bishop=3, Rook=4, Queen=5, King=6. pieceTypeIndex (features.go) remaps
# the five non-king types to a dense 0..4 range; King is never looked up
# in it (ActiveFeatures skips king squares entirely).
PIECE_TYPE_INDEX = {1: 0, 2: 1, 3: 2, 4: 3, 5: 4}
KING_PIECE_TYPE = 6

FEN_PIECE_TYPE = {"p": 1, "n": 2, "b": 3, "r": 4, "q": 5, "k": 6}


def king_bucket(king_file, king_rank):
    """Mirrors features.go's kingBucket: quadrant of the board White's king sits in."""
    bucket = 0
    if king_file >= 4:
        bucket |= 1
    if king_rank >= 4:
        bucket |= 2
    return bucket


def piece_square_feature(square, piece_type, color):
    return square * 5 * 2 + PIECE_TYPE_INDEX[piece_type] * 2 + color


def piece_feature(bucket, square, piece_type, color):
    return bucket * NUM_PIECE_SQUARE_FEATURES + piece_square_feature(square, piece_type, color)


def _append_count_bucket(feats, base, count):
    bucket = min(count, NUM_COUNT_BUCKETS - 1)
    feats.append(base + bucket)
    return base + NUM_COUNT_BUCKETS


class Board:
    """Piece placement decoded from a FEN's first field only. Square
    numbering matches core.Square exactly: rank*8+file, 0=a1 .. 63=h8
    (see core/bitboard.go's Square doc comment and ParseFEN's own
    rank-8-first-in-FEN-but-rank-0-first-internally handling)."""

    def __init__(self, fen):
        self.piece_at = {}
        self.white_king_square = None

        board_field = fen.split()[0]
        ranks = board_field.split("/")
        if len(ranks) != 8:
            raise ValueError(f"FEN board must have 8 ranks: {fen!r}")

        for i, rank_str in enumerate(ranks):
            rank = 7 - i
            file = 0
            for ch in rank_str:
                if ch.isdigit():
                    file += int(ch)
                    continue
                color = BLACK if ch.islower() else WHITE
                piece_type = FEN_PIECE_TYPE[ch.lower()]
                square = rank * 8 + file
                self.piece_at[square] = (piece_type, color)
                if piece_type == KING_PIECE_TYPE and color == WHITE:
                    self.white_king_square = square
                file += 1
            if file != 8:
                raise ValueError(f"rank {rank_str!r} does not sum to 8 files: {fen!r}")

        if self.white_king_square is None:
            raise ValueError(f"FEN has no white king: {fen!r}")


def active_features(
    fen,
    white_hand_size,
    black_hand_size,
    frozen_white,
    frozen_black,
    shielded_white,
    shielded_black,
    fortress_white,
    fortress_black,
):
    """Python equivalent of nnue.ActiveFeatures. Returns the list of active
    feature indices for the given position/card-state summary -- order is
    not meaningful (ActiveFeatures's own order depends on Go's map/loop
    iteration and this port's differs), only the resulting SET matters,
    exactly like a multi-hot input vector's nonzero positions."""
    board = Board(fen)
    king_file, king_rank = board.white_king_square % 8, board.white_king_square // 8
    bucket = king_bucket(king_file, king_rank)

    feats = []
    for square, (piece_type, color) in sorted(board.piece_at.items()):
        if piece_type == KING_PIECE_TYPE:
            continue
        feats.append(piece_feature(bucket, square, piece_type, color))

    base = NUM_CHESS_FEATURES
    base = _append_count_bucket(feats, base, white_hand_size)
    base = _append_count_bucket(feats, base, black_hand_size)
    base = _append_count_bucket(feats, base, frozen_white)
    base = _append_count_bucket(feats, base, frozen_black)
    base = _append_count_bucket(feats, base, shielded_white)
    base = _append_count_bucket(feats, base, shielded_black)
    if fortress_white:
        feats.append(base)
    base += 1
    if fortress_black:
        feats.append(base)
    base += 1

    return feats
